### ISSUE TITLE
1.19.1 resamplers (instead 1.18.2) introduce distortion into output signal

### ISSUE THREAD

**Issue by vmech** [kcat/openal-soft#345]
I typically used cubic resampler (from 44.1kHz to 48kHz) with 1.18.2, it worked flawlessly.
But recently updated soft_oal.dll into 1.19.1 version. And, unfortunately, discovered that new version introduces audible and significant distortion into output signal, with ANY resamplers (cubic, bsinc12, bsinc24).

What happened? How could this happen?

Used: Audigy SE (CA0106-based DSP), Win 10 64-bit, foobar 1.4.4/1.4.6 (no any DSPs), CrystalMixer 1.1.0.6 + soft_oal 1.18.2 / CrystalMixer 1.2.0.0 + soft_oal 1.19.1

**Comment by kcat** [comment]
What kind of distortion are you hearing? Are you able to make a recording of the noise, or a comparison? Does it still happen with the latest master, and is it with a 32-bit lib/app or 64-bit?

**Comment by vmech** [comment]
I hear sound distortion :)

Foobar 1.4.4 32-bit, Crystal Mixer 1.1.0.6/1.2.0.0 32-bit, soft_oal.dll 1.18.2 32-bit/1.19.1 32-bit

Clean spectrum (No alsoft.ini, foobar + Sox (mod) resampler):
[![](https://i.imgur.com/0vdGcaom.jpg)](https://i.imgur.com/0vdGcao.png)
[FLAC](https://www.dropbox.com/s/djlsadmwyghxhlz/Clean_ReSox.flac?dl=0)

A bit dirty but quite acceptable (alsoft.ini >> resampler=cubic, foobar):
[![](https://i.imgur.com/gSF1mZKm.jpg)](https://i.imgur.com/gSF1mZK.png)
[FLAC](https://www.dropbox.com/s/fi5xs7n16gsziqq/1.18.2_Cubic_almost_clean.flac?dl=0)

Unacceptable dirty (alsoft.ini >> resampler=cubic, foobar):
[![](https://i.imgur.com/JFoHD58m.jpg)](https://i.imgur.com/JFoHD58.png)
[FLAC](https://www.dropbox.com/s/hz4percib6x4st3/1.19.1_Cubic_dirty.flac?dl=0)

**Comment by kcat** [comment]
It's notable that 1.18 doesn't have a cubic resampler, requesting it instead gives a sinc4 resampler. 1.19 changed it to a proper cubic resampler, and the main reason for that is performance... the sinc4 resampler was only very marginally faster than bsinc12 but notably worse quality, and the C version of the cubic resampler was faster than the SSE version of the sinc4 resampler.

If it's performance you want, cubic is faster than the bsinc resamplers while still better than linear. If it's quality you're interested in, bsinc12 or even bsinc24 would be better (1.18 only has bsinc, which is similar to bsinc12).

**Comment by vmech** [comment]
Yes, you are right, I did not pay attention to it:
```
AL lib: (II) alc_initconfig: Initializing library v1.18.2-ce607609 v1.18
...
AL lib: (II) GetConfigValue: Found resampler = "cubic"
AL lib: (WW) aluInitMixer: Resampler option "cubic" is deprecated, using sinc4
```
I noticed changes with sounds in STALKER: Dead Air videogame after updating soft_oal.dll from 1.18.2 to 1.19.1. Just I using Foobar and Audacity for checking sounds anomalies.

Now I need some times to do additional tests with bsinc12 resampler for both libraries in game to determine the impact on performance/quality.

But one thing is certain: cubic resampler is completely unusable - on cheapest discrete soundboard with not the most expensive but quality headphones the sound distortion is heard very clearly when resampling by cubic in 1.19.1 library.

**Comment by donizettilorenzo** [comment]
I confirm the issue ;)

**Comment by vmech** [comment]
I got to the computer and completed the quality tests :) There are conclusions:
- Only 2 resamplers have acceptable quality: 1.18.2 Sinc4 and 1.19.1 Bsinc24
- Not one softoal resampler reaches the quality of a specialized resampler for audioplayers (by the way, very fast and high-quality)

Unexpected (for me personally):
- 1.18.2: Sinc4 has better quality than Bsinc
- 1.19.1: Bsinc12 has simply failed in quality

resampler|spectrogram|sample
-|-|:-:
Source|[![](https://i.imgur.com/9IWhy6Cm.jpg)](https://i.imgur.com/9IWhy6C.png)|[FLAC](https://www.dropbox.com/s/giiumd3dplzwoxv/udial.flac?dl=0)
1.18.2 Sinc4|[![](https://i.imgur.com/jGfteOkm.jpg)](https://i.imgur.com/jGfteOk.png)|[FLAC](https://www.dropbox.com/s/unb878shms3bz3i/1.18.2_Sinc4_almost_clean.flac?dl=0)
1.18.2 Bsinc|[![](https://i.imgur.com/B8yx0MIm.jpg)](https://i.imgur.com/B8yx0MI.png)|[FLAC](https://www.dropbox.com/s/ch6nee0pbzkvmin/1.18.2_Bsinc_almost_clean.flac?dl=0)
1.19.1 Cubic|[![](https://i.imgur.com/1SLoTLhm.jpg)](https://i.imgur.com/1SLoTLh.png)|[FLAC](https://www.dropbox.com/s/hz4percib6x4st3/1.19.1_Cubic_dirty.flac?dl=0)
1.19.1 Bsinc12|[![](https://i.imgur.com/Xfln13ym.jpg)](https://i.imgur.com/Xfln13y.png)|[FLAC](https://www.dropbox.com/s/h4eburoze5c2c69/1.19.1_Bsinc12_dirty.flac?dl=0)
1.19.2 Bsinc24|[![](https://i.imgur.com/p0vykcjm.jpg)](https://i.imgur.com/p0vykcj.png)|[FLAC](https://www.dropbox.com/s/58074r8741af98h/1.19.1_Bsinc24_almost_clean.flac?dl=0)
 | | 
Sox (mod)|[![](https://i.imgur.com/2AYtNT8m.jpg)](https://i.imgur.com/2AYtNT8.png)|[FLAC](https://www.dropbox.com/s/djlsadmwyghxhlz/Clean_ReSox.flac?dl=0)

**Comment by kcat** [comment]
That is certainly unexpected, and doesn't quite align with the measurements done a few years ago when these were originally added. 1.18.2's bsinc should vastly outperform sInc4 in quality, with only a minuscule drop in speed. This is the measurements I got leading up to 1.17's release:
![ResamplerPlots](https://user-images.githubusercontent.com/1545516/65723611-9c4a2600-e063-11e9-8458-1cffe18d7ed6.png)
(The Band Limited Sinc Interpolator is 1.17/18's bsinc, and the Polyphase Sinc Resampler is what was in the makehrtf/makemhr tool). 1.19.1's bsinc12 being a bit noisier than 1.18.2's bsinc is expected; for 1.19, I raised the drop-off level from -60dB to -40dB, to help reduce the transition band and keep some higher frequencies. Though I don't know why bsinc itself is that noisy to begin with.

Out of curiosity, how is the resampling being done? The source and outputs all seem to be 44.1khz, and OpenAL Soft is supposed to disable the resampler when the input and output rates match.

Also, can you give a bit of context to the "very fast" specialized resamplers in other players? For reference, OpenAL needs to be able to resample hundreds of sources in real-time, along with various filters and effects, while playing a demanding game, whereas an audio player generally only needs to resample one or two streams while perhaps decoding a video.

**Comment by vmech** [comment]
From 44.1kHz to 48kHz. I added 48 kHz 24-bit final samples to the [previous post](https://github.com/kcat/openal-soft/issues/345#issuecomment-535593338).

**Comment by vmech** [comment]
Testing with [Source FLAC](https://www.dropbox.com/s/giiumd3dplzwoxv/udial.flac?dl=0)

**1 thread**:
```
System:
  CPU: AMD Phenom(tm) II X4 965 Processor, features: 3DNow!ex MMX SSE SSE2 SSE3
  App: foobar2000 v1.4.4
Settings:
  High priority: yes
  Buffer entire file into memory: yes
  Warm-up: yes
  Passes: 100
  Threads: 1
  Postprocessing: Resampler (SoX) mod
Stats by codec:
  FLAC: 131.565x realtime
Total:
  Decoded length: 4:35.002
  Opening time: 0:00.013
  Decoding time: 0:02.077
  Speed (x realtime): 131.565
```
**4 threads**:
```
System:
  CPU: AMD Phenom(tm) II X4 965 Processor, features: 3DNow!ex MMX SSE SSE2 SSE3
  App: foobar2000 v1.4.4
Settings:
  High priority: yes
  Buffer entire file into memory: yes
  Warm-up: yes
  Passes: 100
  Threads: 4
  Postprocessing: Resampler (SoX) mod
Stats by codec:
  FLAC: 531.850x realtime
Total:
  Decoded length: 18:20.009
  Opening time: 0:00.054
  Decoding time: 0:08.219
  Speed (x realtime): 531.850
```

**Comment by vmech** [comment]
Hey @kcat I have a question. What mean `AL lib: (II) alc_initconfig: Initialized backend "wasapi"` `AL lib: (II) alc_initconfig: Added "wasapi" for playback` `AL lib: (II) alc_initconfig: Added "wasapi" for capture` ?
After processing in softoal the sound is fed to the wasapi input and goes through the entire standard audio stack (CPT -> APO1 -> APO2 -> Mixer -> APO3 -> KST) with all its inherent distortions, and only after that it gets into the driver -> soundboard?

**Comment by kcat** [comment]
> But, I suppose, you should perfectly understand that software processing by CPU «do not have to grab the stars from the sky».

I'm not sure what you mean. I'm not familiar with that saying, "do not have to grab the stars from the sky".

> After processing in softoal sound is fed to wasapi instance and goes through the entire standard windows audiostack (CPT -> APO1 -> APO2 -> Mixer -> APO3 -> KST) with all its inherent distortions, and only after that it gets into the driver -> soundboard?

OpenAL Soft's sound mix is fed to a WASAPI/MMDeviceAPI stream, which is the normal output method on Windows. Short of using something like hardware driver interfaces or ASIO, that's how all audio is played on Windows since Vista. Even DirectSound and Windows Multimedia use WASAPI for audio playback and capture.

**Comment by vmech** [comment]
> OpenAL Soft's sound mix is fed to a WASAPI/MMDeviceAPI stream, which is the normal output method on Windows. Short of using something like hardware driver interfaces or ASIO, that's how all audio is played on Windows since Vista. Even DirectSound and Windows Multimedia use WASAPI for audio playback and capture.

I asked, because there is such a scheme. It shows that the true OpenAL has no any contacts with WASAPI instance.
[![](https://i.imgur.com/leJcjxdm.jpg)](https://i.imgur.com/leJcjxd.png)

But thank you for the answer. I found out everything I wanted. Programs like ASIO:Crystal Mixer have nothing to do with true ASIO, and do not produce any of its software “reincarnation”. The only way to get clear sound from the player is through WASAPI Exclusive or Kernel streaming.

**Comment by kcat** [comment]
By "true OpenAL" I guess you mean Creative's hardware driver, ct_oal.dll. That can of course skip Windows' audio stack since it can issue commands to the hardware through custom interfaces in Creative's driver. A normal application would have to be specially coded to look for and use that specific driver to do the same.

ASIO is a special case in that its entire purpose is to provide a low-latency common interface that outputs direct to hardware, for audio production work. Its developers likely work with device manufacturers to utilize their drivers or reverse engineer it themselves.

WASAPI Exclusive mode has issues in that it's exclusive. Other apps and the system wouldn't be able to play anything on the device while OpenAL Soft is playing.

By the way, what is your device configuration in Windows like? Are you using headphones or surround sound? Can you provide the full trace log?

**Comment by vmech** [comment]
> By "true OpenAL" I guess you mean Creative's hardware OpenAL driver, ct_oal.dll. That can of course skip Windows' audio stack since it can issue commands to the hardware through custom interfaces in Creative's device driver. A normal application would have to be specially coded to look for and use that specific driver to do the same.

Creative software-based driver has that one named as sens_oal.dll which can hardlinked to ct_oal.dll too.

[![](https://i.imgur.com/6s6rH3Zm.jpg)](https://i.imgur.com/6s6rH3Z.png)

> ASIO is a special case in that its entire purpose is to provide a low-latency common interface that outputs direct to hardware, for audio production work. Its developers likely work with device manufacturers to utilize their drivers or reverse engineer it themselves.

I know, ASIO is nontrivially complicated thing.

> WASAPI Exclusive mode has issues in that it's exclusive. Other apps and the system wouldn't be able to play anything on the device while OpenAL Soft is playing.

But can «Exclusive» mean something else? :)

>By the way, what is your device configuration in Windows like? Are you using headphones or surround sound? Can you provide the full trace log?

Generally I use headphones. But in Creative control panel I can change Windows speakers configuration on the fly, and driver rerouting and mixing them into headphones through soundboard DSP.

[![](https://i.imgur.com/eygi6Ojm.jpg)](https://i.imgur.com/eygi6Oj.png)

2.0/2.1 over Headphones trace log:
```
AL lib: (II) alc_initconfig: Initializing library v1.18.2-ce607609 v1.18
AL lib: (II) alc_initconfig: Supported backends: mmdevapi, dsound, winmm, null, wave
AL lib: (II) ReadALConfig: Loading config C:\Users\sCreamER\AppData\Roaming\alsoft.ini...
AL lib: (II) LoadConfigFromFile: found 'output-limiter' = 'false'
AL lib: (II) GetProcPath: Got: F:\GAMES\STALKER-DA
AL lib: (II) ReadALConfig: Loading config F:\GAMES\STALKER-DA\alsoft.ini...
AL lib: (II) LoadConfigFromFile: found 'resampler' = 'sinc4'
AL lib: (II) GetConfigValue: Key disable-cpu-exts not found
AL lib: (II) FillCPUCaps: Detected max CPUID function: 0x5 (ext. 0x8000001b)
AL lib: (II) FillCPUCaps: Vendor ID: "AuthenticAMD"
AL lib: (II) FillCPUCaps: Name: "AMD Phenom(tm) II X4 965 Processor"
AL lib: (II) FillCPUCaps: Extensions: +SSE +SSE2 +SSE3 -SSE4.1
AL lib: (II) GetConfigValue: Key rt-prio not found
AL lib: (II) GetConfigValue: Found resampler = "sinc4"
AL lib: (II) GetConfigValue: Key trap-al-error not found
AL lib: (II) GetConfigValue: Key trap-alc-error not found
AL lib: (II) GetConfigValue: Key reverb/boost not found
AL lib: (II) GetConfigValue: Key reverb/emulate-eax not found
AL lib: (II) GetConfigValue: Key drivers not found
AL lib: (II) ALCmmdevProxy_messageHandler: Starting message thread
AL lib: (II) ALCmmdevProxy_messageHandler: Message thread initialization complete
AL lib: (II) ALCmmdevProxy_messageHandler: Starting message loop
AL lib: (II) alc_initconfig: Initialized backend "mmdevapi"
AL lib: (II) alc_initconfig: Added "mmdevapi" for playback
AL lib: (II) alc_initconfig: Added "mmdevapi" for capture
AL lib: (II) GetConfigValue: Key excludefx not found
AL lib: (II) GetConfigValue: Key default-reverb not found
AL lib: (II) GetConfigValue: Key channels not found
AL lib: (II) GetConfigValue: Key sample-type not found
AL lib: (II) GetConfigValue: Key frequency not found
AL lib: (II) GetConfigValue: Key periods not found
AL lib: (II) GetConfigValue: Key period_size not found
AL lib: (II) GetConfigValue: Key sources not found
AL lib: (II) GetConfigValue: Key slots not found
AL lib: (II) GetConfigValue: Key sends not found
AL lib: (II) ALCmmdevProxy_messageHandler: Got message "Open Device" (0x0400, lparam=00EBAE10, wparam=0019F6DC)
AL lib: (II) GetConfigValue: Key ambi-format not found
AL lib: (II) alcOpenDevice: Created device 06220050, "OpenAL Soft on Speakers (SB Audigy)"
AL lib: (II) GetConfigValue: Key hrtf not found
AL lib: (II) UpdateDeviceParams: Pre-reset: Stereo, Float, 44100hz, 1024 update size x3
AL lib: (II) ALCmmdevProxy_messageHandler: Got message "Reset Device" (0x0401, lparam=00EBAE10, wparam=0019F73C)
AL lib: (II) UpdateDeviceParams: Post-reset: Stereo, Float, 48000hz, 960 update size x3
AL lib: (II) GetConfigValue: Key stereo-mode not found
AL lib: (II) aluInitRenderer: HRTF disabled
AL lib: (II) GetConfigValue: Key cf_level not found
AL lib: (II) aluInitRenderer: BS2B disabled
AL lib: (II) GetConfigValue: Key stereo-encoding not found
AL lib: (II) aluInitRenderer: UHJ disabled
AL lib: (II) UpdateDeviceParams: Channel config, Dry: 2, FOA: 0, Real: 0
AL lib: (II) UpdateDeviceParams: Allocating 2 channels, 16384 bytes
AL lib: (II) UpdateDeviceParams: Max sources: 256 (255 + 1), effect slots: 64, sends: 2
AL lib: (II) GetConfigValue: Key dither not found
AL lib: (II) GetConfigValue: Key dither-depth not found
AL lib: (II) UpdateDeviceParams: Dithering disabled
AL lib: (II) GetConfigValue: Found output-limiter = "false"
AL lib: (II) UpdateDeviceParams: Output limiter disabled
AL lib: (II) ALCmmdevProxy_messageHandler: Got message "Start Device" (0x0402, lparam=00EBAE10, wparam=0019F73C)
AL lib: (II) GetConfigValue: Key volume-adjust not found
AL lib: (II) alcCreateContext: Created context 00EBC1F0
AL lib: (II) ALCmmdevProxy_messageHandler: Got message "Stop Device" (0x0403, lparam=00EBAE10, wparam=0019F814)
AL lib: (II) FreeContext: 00EBC1F0
AL lib: (II) FreeContext: Freed unapplied listener update 00EBC5C0
AL lib: (II) FreeContext: Freed 0 listener property objects
AL lib: (II) FreeDevice: 06220050
AL lib: (II) ALCmmdevProxy_messageHandler: Got message "Close Device" (0x0404, lparam=00EBAE10, wparam=0019F7DC)
AL lib: (II) GetConfigValue: Key channels not found
AL lib: (II) GetConfigValue: Key sample-type not found
AL lib: (II) GetConfigValue: Key frequency not found
AL lib: (II) GetConfigValue: Key periods not found
AL lib: (II) GetConfigValue: Key period_size not found
AL lib: (II) GetConfigValue: Key sources not found
AL lib: (II) GetConfigValue: Key slots not found
AL lib: (II) GetConfigValue: Key sends not found
AL lib: (II) ALCmmdevProxy_messageHandler: Got message "Open Device" (0x0400, lparam=00EBAE10, wparam=0019F814)
AL lib: (II) GetConfigValue: Key ambi-format not found
AL lib: (II) alcOpenDevice: Created device 06220050, "OpenAL Soft on Speakers (SB Audigy)"
AL lib: (II) GetConfigValue: Key hrtf not found
AL lib: (II) UpdateDeviceParams: Pre-reset: Stereo, Float, 44100hz, 1024 update size x3
AL lib: (II) ALCmmdevProxy_messageHandler: Got message "Reset Device" (0x0401, lparam=00EBAE10, wparam=0019F870)
AL lib: (II) UpdateDeviceParams: Post-reset: Stereo, Float, 48000hz, 960 update size x3
AL lib: (II) GetConfigValue: Key stereo-mode not found
AL lib: (II) aluInitRenderer: HRTF disabled
AL lib: (II) GetConfigValue: Key cf_level not found
AL lib: (II) aluInitRenderer: BS2B disabled
AL lib: (II) GetConfigValue: Key stereo-encoding not found
AL lib: (II) aluInitRenderer: UHJ disabled
AL lib: (II) UpdateDeviceParams: Channel config, Dry: 2, FOA: 0, Real: 0
AL lib: (II) UpdateDeviceParams: Allocating 2 channels, 16384 bytes
AL lib: (II) UpdateDeviceParams: Max sources: 256 (255 + 1), effect slots: 64, sends: 2
AL lib: (II) GetConfigValue: Key dither not found
AL lib: (II) GetConfigValue: Key dither-depth not found
AL lib: (II) UpdateDeviceParams: Dithering disabled
AL lib: (II) GetConfigValue: Found output-limiter = "false"
AL lib: (II) UpdateDeviceParams: Output limiter disabled
AL lib: (II) ALCmmdevProxy_messageHandler: Got message "Start Device" (0x0402, lparam=00EBAE10, wparam=0019F870)
AL lib: (II) GetConfigValue: Key volume-adjust not found
AL lib: (II) alcCreateContext: Created context 00EBDE60
AL lib: (II) ALCmmdevProxy_messageHandler: Got message "Stop Device" (0x0403, lparam=00EBAE10, wparam=0019F990)
AL lib: (II) FreeContext: 00EBDE60
AL lib: (II) FreeContext: Freed 2 listener property objects
AL lib: (II) FreeDevice: 06220050
AL lib: (II) ALCmmdevProxy_messageHandler: Got message "Close Device" (0x0404, lparam=00EBAE10, wparam=0019F954)
```
5.1 over Headphones trace log:
```
AL lib: (II) alc_initconfig: Initializing library v1.18.2-ce607609 v1.18
AL lib: (II) alc_initconfig: Supported backends: mmdevapi, dsound, winmm, null, wave
AL lib: (II) ReadALConfig: Loading config C:\Users\sCreamER\AppData\Roaming\alsoft.ini...
AL lib: (II) LoadConfigFromFile: found 'output-limiter' = 'false'
AL lib: (II) GetProcPath: Got: F:\GAMES\STALKER-DA
AL lib: (II) ReadALConfig: Loading config F:\GAMES\STALKER-DA\alsoft.ini...
AL lib: (II) LoadConfigFromFile: found 'resampler' = 'sinc4'
AL lib: (II) GetConfigValue: Key disable-cpu-exts not found
AL lib: (II) FillCPUCaps: Detected max CPUID function: 0x5 (ext. 0x8000001b)
AL lib: (II) FillCPUCaps: Vendor ID: "AuthenticAMD"
AL lib: (II) FillCPUCaps: Name: "AMD Phenom(tm) II X4 965 Processor"
AL lib: (II) FillCPUCaps: Extensions: +SSE +SSE2 +SSE3 -SSE4.1
AL lib: (II) GetConfigValue: Key rt-prio not found
AL lib: (II) GetConfigValue: Found resampler = "sinc4"
AL lib: (II) GetConfigValue: Key trap-al-error not found
AL lib: (II) GetConfigValue: Key trap-alc-error not found
AL lib: (II) GetConfigValue: Key reverb/boost not found
AL lib: (II) GetConfigValue: Key reverb/emulate-eax not found
AL lib: (II) GetConfigValue: Key drivers not found
AL lib: (II) ALCmmdevProxy_messageHandler: Starting message thread
AL lib: (II) ALCmmdevProxy_messageHandler: Message thread initialization complete
AL lib: (II) ALCmmdevProxy_messageHandler: Starting message loop
AL lib: (II) alc_initconfig: Initialized backend "mmdevapi"
AL lib: (II) alc_initconfig: Added "mmdevapi" for playback
AL lib: (II) alc_initconfig: Added "mmdevapi" for capture
AL lib: (II) GetConfigValue: Key excludefx not found
AL lib: (II) GetConfigValue: Key default-reverb not found
AL lib: (II) GetConfigValue: Key channels not found
AL lib: (II) GetConfigValue: Key sample-type not found
AL lib: (II) GetConfigValue: Key frequency not found
AL lib: (II) GetConfigValue: Key periods not found
AL lib: (II) GetConfigValue: Key period_size not found
AL lib: (II) GetConfigValue: Key sources not found
AL lib: (II) GetConfigValue: Key slots not found
AL lib: (II) GetConfigValue: Key sends not found
AL lib: (II) ALCmmdevProxy_messageHandler: Got message "Open Device" (0x0400, lparam=00F8AE10, wparam=0019F6DC)
AL lib: (II) GetConfigValue: Key ambi-format not found
AL lib: (II) alcOpenDevice: Created device 06180050, "OpenAL Soft on Speakers (SB Audigy)"
AL lib: (II) GetConfigValue: Key hrtf not found
AL lib: (II) UpdateDeviceParams: Pre-reset: Stereo, Float, 44100hz, 1024 update size x3
AL lib: (II) ALCmmdevProxy_messageHandler: Got message "Reset Device" (0x0401, lparam=00F8AE10, wparam=0019F73C)
AL lib: (II) UpdateDeviceParams: Post-reset: 5.1 Surround (Rear), Float, 48000hz, 960 update size x3
AL lib: (II) GetConfigValue: Key decoder/surround51 not found
AL lib: (II) UpdateDeviceParams: Channel config, Dry: 6, FOA: 0, Real: 0
AL lib: (II) UpdateDeviceParams: Allocating 6 channels, 49152 bytes
AL lib: (II) UpdateDeviceParams: Max sources: 256 (255 + 1), effect slots: 64, sends: 2
AL lib: (II) GetConfigValue: Key dither not found
AL lib: (II) GetConfigValue: Key dither-depth not found
AL lib: (II) UpdateDeviceParams: Dithering disabled
AL lib: (II) GetConfigValue: Found output-limiter = "false"
AL lib: (II) UpdateDeviceParams: Output limiter disabled
AL lib: (II) ALCmmdevProxy_messageHandler: Got message "Start Device" (0x0402, lparam=00F8AE10, wparam=0019F73C)
AL lib: (II) GetConfigValue: Key volume-adjust not found
AL lib: (II) alcCreateContext: Created context 00F8C1F0
AL lib: (II) ALCmmdevProxy_messageHandler: Got message "Stop Device" (0x0403, lparam=00F8AE10, wparam=0019F814)
AL lib: (II) FreeContext: 00F8C1F0
AL lib: (II) FreeContext: Freed unapplied listener update 00F8C5C0
AL lib: (II) FreeContext: Freed 0 listener property objects
AL lib: (II) FreeDevice: 06180050
AL lib: (II) ALCmmdevProxy_messageHandler: Got message "Close Device" (0x0404, lparam=00F8AE10, wparam=0019F7DC)
AL lib: (II) GetConfigValue: Key channels not found
AL lib: (II) GetConfigValue: Key sample-type not found
AL lib: (II) GetConfigValue: Key frequency not found
AL lib: (II) GetConfigValue: Key periods not found
AL lib: (II) GetConfigValue: Key period_size not found
AL lib: (II) GetConfigValue: Key sources not found
AL lib: (II) GetConfigValue: Key slots not found
AL lib: (II) GetConfigValue: Key sends not found
AL lib: (II) ALCmmdevProxy_messageHandler: Got message "Open Device" (0x0400, lparam=00F8AE10, wparam=0019F814)
AL lib: (II) GetConfigValue: Key ambi-format not found
AL lib: (II) alcOpenDevice: Created device 06180050, "OpenAL Soft on Speakers (SB Audigy)"
AL lib: (II) GetConfigValue: Key hrtf not found
AL lib: (II) UpdateDeviceParams: Pre-reset: Stereo, Float, 44100hz, 1024 update size x3
AL lib: (II) ALCmmdevProxy_messageHandler: Got message "Reset Device" (0x0401, lparam=00F8AE10, wparam=0019F870)
AL lib: (II) UpdateDeviceParams: Post-reset: 5.1 Surround (Rear), Float, 48000hz, 960 update size x3
AL lib: (II) GetConfigValue: Key decoder/surround51 not found
AL lib: (II) UpdateDeviceParams: Channel config, Dry: 6, FOA: 0, Real: 0
AL lib: (II) UpdateDeviceParams: Allocating 6 channels, 49152 bytes
AL lib: (II) UpdateDeviceParams: Max sources: 256 (255 + 1), effect slots: 64, sends: 2
AL lib: (II) GetConfigValue: Key dither not found
AL lib: (II) GetConfigValue: Key dither-depth not found
AL lib: (II) UpdateDeviceParams: Dithering disabled
AL lib: (II) GetConfigValue: Found output-limiter = "false"
AL lib: (II) UpdateDeviceParams: Output limiter disabled
AL lib: (II) ALCmmdevProxy_messageHandler: Got message "Start Device" (0x0402, lparam=00F8AE10, wparam=0019F870)
AL lib: (II) GetConfigValue: Key volume-adjust not found
AL lib: (II) alcCreateContext: Created context 00F8DE60
AL lib: (II) ALCmmdevProxy_messageHandler: Got message "Stop Device" (0x0403, lparam=00F8AE10, wparam=0019F990)
AL lib: (II) FreeContext: 00F8DE60
AL lib: (II) FreeContext: Freed 2 listener property objects
AL lib: (II) FreeDevice: 06180050
AL lib: (II) ALCmmdevProxy_messageHandler: Got message "Close Device" (0x0404, lparam=00F8AE10, wparam=0019F954)
```
When using 7.1 over headphones, the same thing happens, but 7.1 Surround is detected and 8 channels are allocated.

**Comment by kcat** [comment]
> Creative software-based driver has that one named as sens_oal.dll which can hardlinked to ct_oal.dll too.

All of Creative's drivers (except wrap_oal.dll which specifically uses DirectSound or WinMM) can be coded to look for and use their hardware driver to avoid the Windows audio stack. As long as the device is capable of hardware mixing, it won't need to block other apps from playing sounds (though it would mean Windows' application volume control won't apply).

> But can *Exclusive* mean something else? :)

In the context of WASAPI, it means taking exclusive control of the device, allowing the app to reconfigure the hardware parameters but preventing other apps from using the device at the same time.

> OpenAL Soft connects via WASAPI Shared, so this does not stop other programs from playing sounds.
> But in this mode sounds passes through all audiostack handlers (APOs), and at output it receives all distortions they introduce.

An APO (Audio Processing Object) is simply a kind of plugin, which can do something to the stream prior to mixing or prior to output to the device. They don't inherently add distortion unless they do something that adds noise. For instance, an APO may apply the overall process volume prior to mixing, which simply scales all samples by a fixed amount (no distortion, except for a little extremely low level quantization noise). Another may add a compressor/limiter prior to the device output, which reduces the volume around samples that exceed the output range (and effectively does nothing when samples don't exceed the output range). Any distortion would have to be a result of something a bit more specific than the audiostack/APOs in general.

Anyway, I've found that doubling the number of phase steps used by the bsinc resamplers helps get rid of a couple of the worst distortion peaks, at least for 44.1khz to 48khz resampling. This doubles the size of the filter tables, but shouldn't impact the cache too badly. I've also readjusted the bsinc12 resampler to have a -60dB cutoff again, reducing the overall noise level at the cost of more high-frequency muffling.
